src/telemetry/hardware_info.rs

```diff
@@ -18,38 +18,139 @@
-use astarte_device_sdk::types::AstarteType;
+use crate::data::{publish, Publisher};
 use procfs::{CpuInfo, Meminfo, ProcResult};
-use std::collections::HashMap;
+use serde::Deserialize;
+use tracing::{debug, error};
 
-/// get structured data for `io.edgehog.devicemanager.HardwareInfo` interface
-pub fn get_hardware_info() -> Result<HashMap<String, AstarteType>, DeviceManagerError> {
-    let mut ret: HashMap<String, AstarteType> = HashMap::new();
+const INTERFACE: &str = "io.edgehog.devicemanager.HardwareInfo";
 
-    let architecture = get_machine_architecture();
-    ret.insert("/cpu/architecture".to_owned(), architecture.into());
+#[derive(Debug, Default, Deserialize)]
+pub struct HardwareInfo {
+    pub cpu: Cpu,
+    pub mem: Mem,
+}
+
+impl HardwareInfo {
+    pub async fn read() -> Self {
+        let cpu = Cpu::read().await;
+        let mem = Mem::read().await;
 
-    let cpuinfo = get_cpu_info()?;
-    if let Some(f) = cpuinfo.fields.get("model") {
-        ret.insert("/cpu/model".to_owned(), f.clone().into());
+        HardwareInfo { cpu, mem }
     }
 
-    if let Some(f) = cpuinfo.fields.get("model name") {
-        ret.insert("/cpu/modelName".to_owned(), f.clone().into());
+    /// get structured data for `io.edgehog.devicemanager.HardwareInfo` interface
+    pub async fn send<T>(self, client: &T)
+    where
+        T: Publisher,
+    {
+        self.cpu.send(client).await;
+        self.mem.send(client).await;
     }
+}
+
+#[derive(Debug, Default, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Cpu {
+    pub architecture: String,
+    pub model: Option<String>,
+    pub model_name: Option<String>,
+    pub vendor: Option<String>,
+}
+
+impl Cpu {
+    async fn read() -> Self {
+        let mut cpu = Cpu {
+            architecture: get_machine_architecture(),
+            ..Default::default()
+        };
+        match get_cpu_info() {
+            Ok(mut cpu_info) => {
+                cpu.model = cpu_info.fields.remove("model");
+                cpu.model_name = cpu_info.fields.remove("model name");
+                cpu.vendor = cpu_info.fields.remove("vendor_id");
+            }
+            Err(err) => {
+                error!(
+                    "couldn't get the cpu info: {}",
+                    stable_eyre::Report::new(err)
+                );
+            }
+        }
 
-    if let Some(f) = cpuinfo.fields.get("vendor_id") {
-        ret.insert("/cpu/vendor".to_owned(), f.clone().into());
+        cpu
     }
 
-    let meminfo = get_meminfo()?;
-    ret.insert(
-        "/mem/totalBytes".to_owned(),
-        (meminfo.mem_total as i64).into(),
-    );
+    async fn send<T>(self, client: &T)
+    where
+        T: Publisher,
+    {
+        publish(client, INTERFACE, "/cpu/architecture", self.architecture).await;
+
+        if let Some(model) = self.model {
+            publish(client, INTERFACE, "/cpu/model", model).await;
+        } else {
+            debug!("missing cpu model");
+        }
```

Comment:
In `system_info.rs`, `os_release.rs` and other files where you perform the `send` operation you don't put any debug info. Of course this is not an error, still I would do similar things the same way (either use match or if-let-else and either decide to debug or not)

Reply:
I think it's the other way around. The system_info didn't have it, now I added it.

Comment if you find other that you think are missing.